vernac/vernacentries.ml

```diff
@@ -2352,350 +2347,361 @@ let vernac_interp_phase c ~pstate =
-    with_module_locality ~atts vernac_declare_scope sc;
-    pstate
+    VtDefault(fun () -> with_module_locality ~atts vernac_declare_scope sc)
   | VernacDelimiters (sc,lr) ->
-    with_module_locality ~atts vernac_delimiters sc lr;
-    pstate
+    VtDefault(fun () -> with_module_locality ~atts vernac_delimiters sc lr)
   | VernacBindScope (sc,rl) ->
-    with_module_locality ~atts vernac_bind_scope sc rl;
-    pstate
+    VtDefault(fun () -> with_module_locality ~atts vernac_bind_scope sc rl)
   | VernacOpenCloseScope (b, s) ->
-    with_section_locality ~atts vernac_open_close_scope (b,s);
-    pstate
+    VtDefault(fun () -> with_section_locality ~atts vernac_open_close_scope (b,s))
   | VernacInfix (mv,qid,sc) ->
-    with_module_locality ~atts vernac_infix mv qid sc;
-    pstate
+    VtDefault(fun () -> with_module_locality ~atts vernac_infix mv qid sc)
   | VernacNotation (c,infpl,sc) ->
-    with_module_locality ~atts vernac_notation c infpl sc;
-    pstate
+    VtDefault(fun () -> with_module_locality ~atts vernac_notation c infpl sc)
   | VernacNotationAddFormat(n,k,v) ->
-    unsupported_attributes atts;
-    Metasyntax.add_notation_extra_printing_rule n k v;
-    pstate
+    VtDefault(fun () ->
+        unsupported_attributes atts;
+        Metasyntax.add_notation_extra_printing_rule n k v)
   | VernacDeclareCustomEntry s ->
-    with_module_locality ~atts vernac_custom_entry s;
-    pstate
+    VtDefault(fun () -> with_module_locality ~atts vernac_custom_entry s)
 
   (* Gallina *)
-  | VernacDefinition (dk,lid,ProveBody(bl,t)) ->
-    with_open_proof ~pstate (with_def_attributes ~atts vernac_definition_interactive dk lid bl t)
-  | VernacDefinition (dk,lid,DefineBody(bl,red_option,c,typ_opt)) ->
-    with_read_proof ~pstate (with_def_attributes ~atts vernac_definition dk lid bl red_option c typ_opt)
+
+  | VernacDefinition (discharge,lid,DefineBody (bl,red_option,c,typ)) ->
+    VtDefault (fun () ->
+      with_def_attributes ~atts
+       vernac_definition discharge lid bl red_option c typ)
+  | VernacDefinition (discharge,lid,ProveBody(bl,typ)) ->
+    VtOpenProof(fun () ->
+      with_def_attributes ~atts
+       vernac_definition_interactive discharge lid bl typ)
+
   | VernacStartTheoremProof (k,l) ->
-    with_open_proof_simple ~pstate (with_def_attributes ~atts vernac_start_proof k l)
-  | VernacEndProof e ->
-    unsupported_attributes atts;
-    vernac_end_proof ?proof ?pstate e
+    VtOpenProof(fun () -> with_def_attributes ~atts vernac_start_proof k l)
   | VernacExactProof c ->
-    unsupported_attributes atts;
-    vernac_require_open_proof ~pstate (vernac_exact_proof c)
+    VtCloseProof(fun ~pstate ->
+        unsupported_attributes atts;
+        vernac_exact_proof ~pstate c)
+
+  | VernacDefineModule (export,lid,bl,mtys,mexprl) ->
+    let i () =
+      unsupported_attributes atts;
+      vernac_define_module export lid bl mtys mexprl in
+    if List.is_empty mexprl then VtNoProof i else VtDefault i
```

Comment:
Can you put a comment `(* XXX eventually we should enforce NoProof in all cases *)`?

Reply:
Sure, one sec.

Reply:
Should we tho?

Reply:
I could imagine having local modules in proofs to be useful to a few people.

Reply:
>I could imagine having local modules in proofs to be useful to a few people.

Then we can be VtDefault in all cases, but it makes no sense to allow `Module M := Foo Bar.` but not `Module M := Foo.` imo

Reply:
Ok, I've placed a comment that should work for now.

Reply:
> Then we can be VtDefault in all cases, but it makes no sense to allow `Module M := Foo Bar.` but not `Module M := Foo.` imo

Yeah this deserves some thinking.